### PR TITLE
use openAppSecrets instead of list+Open secret in DataSourceVaultSecr…

### PR DESCRIPTION
**Pull request by hghaf099**
### :hammer_and_wrench: Description

Addresses: https://hashicorp.atlassian.net/browse/VAULT-26495
We currently call ListAppSecrets, iterate over each secret and individually invoke OpenAppSecret [here](https://github.com/hashicorp/terraform-provider-hcp/blob/07e943eda98e351035443487165847c84757de5f/internal/provider/vaultsecrets/data_source_vault_secrets_app.go#L101-L117). This is both inefficient for us and causes additional API usage for the clients.

I think switching to OpenAppSecrets for this data source would be best.

### :building_construction: Acceptance tests

- [ ] Are there any feature flags that are required to use this functionality?
- [ ] Have you added an acceptance test for the functionality being added?
- [x] Have you run the acceptance tests on this branch?

Output from acceptance testing:

```sh
$ make testacc TESTARGS='-run=TestAcc_dataSourceVaultSecretsApp'

...

=== RUN   TestAcc_dataSourceVaultSecretsAppMigration
--- PASS: TestAcc_dataSourceVaultSecretsAppMigration (12.94s)
=== RUN   TestAcc_dataSourceVaultSecretsApp
    data_source_vault_secrets_app_test.go:97: secrets:  secret_one  =  hey, this is version 1!
    data_source_vault_secrets_app_test.go:97: secrets:  secret_two  =  hey, this is version 2!
--- PASS: TestAcc_dataSourceVaultSecretsApp (6.39s)
PASS
ok      github.com/hashicorp/terraform-provider-hcp/internal/provider/vaultsecrets      20.318s
testing: warning: no tests to run
PASS
```
To test this, I also created the following helper function:
```

func openTestAppSecret(t *testing.T, appName string) {
	t.Helper()

	client := acctest.HCPClients(t)

	loc := &sharedmodels.HashicorpCloudLocationLocation{
		OrganizationID: client.Config.OrganizationID,
		ProjectID:      client.Config.ProjectID,
	}

	secrets, err := clients.OpenVaultSecretsAppSecrets(context.Background(), client, loc, appName)
	if err != nil {
		t.Fatal(err)
	}
	if len(secrets) != 2 {
		t.Fatal("failed to read the secrets")
	}
	for _, s := range secrets {
		t.Log("secrets: ", s.Name, " = ", s.Version.Value)
	}

}
```
Added the above function to `TestAcc_dataSourceVaultSecretsApp` test on line 97 (which is after the test creates two secrets). The test passes and generates the following DD log:

image

### CODE REVIEW

**Review by codergs — COMMENTED** [review]
Great, this looks good me!

**Comment by codergs** [comment]
Not seeing the test `openTestAppSecret ` you mentioned in the comments.

**Comment by hghaf099** [comment]
> Not seeing the test `openTestAppSecret ` you mentioned in the comments.

I provided a snippet of that. I just used that for local testing.

**Comment by codergs** [comment]
> > Not seeing the test `openTestAppSecret ` you mentioned in the comments.
> 
> 
> 
> I provided a snippet of that. I just used that for local testing. 

If you want this test persisted, then we should be able to extend our acceptance tests easily.

**Comment by hghaf099** [comment]
> > > Not seeing the test `openTestAppSecret ` you mentioned in the comments.
> > 
> > 
> > I provided a snippet of that. I just used that for local testing.
> 
> If you want this test persisted, then we should be able to extend our acceptance tests easily.

So, the `TestAcc_dataSourceVaultSecretsApp` test actually practices the new code path. So, [this data source](https://registry.terraform.io/providers/hashicorp/hcp/latest/docs/data-sources/vault_secrets_app) runs the new code for which I have posted the DD log screen shot. The code snippet I provided was just for own sake as part of double assurance. But, that is really not necessary in my view.

**Comment by codergs** [comment]
> > > > Not seeing the test `openTestAppSecret ` you mentioned in the comments.
> 
> > > 
> 
> > > 
> 
> > > I provided a snippet of that. I just used that for local testing.
> 
> > 
> 
> > If you want this test persisted, then we should be able to extend our acceptance tests easily.
> 
> 
> 
> So, the `TestAcc_dataSourceVaultSecretsApp` test actually practices the new code path. So, [this data source](https://registry.terraform.io/providers/hashicorp/hcp/latest/docs/data-sources/vault_secrets_app) runs the new code for which I have posted the DD log screen shot. The code snippet I provided was just for own sake as part of double assurance. But, that is really not necessary in my view. 

Ahh, understood. Existing test is fine then.

**Comment by codergs** [comment]
Can't approve as it's in draft.

**Review by codergs — APPROVED** [review]

File: internal/clients/response.go [hunk 1]
```diff
@@ -31,3 +31,8 @@ func IsResponseCodeInternalError(erro error) bool {
 		return strings.Contains(erro.Error(), fmt.Sprintf("[%d]", http.StatusInternalServerError))
 	}
 }
+
+type ErrorWithCode interface {
```
Comment (manish-hashicorp) [inline internal/clients/response.go:35]: Please consider documenting this since this is exported.

**Review by manish-hashicorp — APPROVED** [review]
Rubber stamping

File: internal/clients/response.go (hunk 1 above)
Comment (hghaf099) [reply internal/clients/response.go:37]: fixed in https://github.com/hashicorp/terraform-provider-hcp/pull/826/commits/aa310a3a72deb8d82363513b328e2f666d76598a

**Review by nywilken — APPROVED** [review]
LGTM

**Pull request merged**